File: Leetcode/PascalsTriangle.cpp
class Solution {
public:
    vector<vector<int>> generate(int numRows) {
        vector< vector<int> > ans;
        if(numRows == 0) return ans;
        
        vector<int> a, b;
        a.push_back(1); 
        ans.push_back(a);
        
        for(int i = 1; i < numRows; ++i){
            b.clear();
            b.push_back(1);
            for(int j = 1; j < a.size(); ++j){
                b.push_back(a[j-1] + a[j]);
            }
            b.push_back(1);
            swap(a, b);
            ans.push_back(a);
        }
        return ans;
        
    }
};

File: Leetcode/PascalsTriangle.rs
// https://leetcode.com/problems/pascals-triangle

struct Solution {}

impl Solution {
    pub fn generate(num_rows: i32) -> Vec<Vec<i32>> {
        let num_rows = num_rows as usize;
        let mut ans: Vec<Vec<i32>> = vec![];
        ans.push(vec![1]);

        for i in 1..num_rows {
            let prev = &ans[i - 1];
            let mut cur = vec![1];
            for j in 0..prev.len() {
                let v = prev[j] + prev.get(j + 1).unwrap_or(&0);
                cur.push(v);
            }
            ans.push(cur);
        }

        ans
    }
}
